### PR TITLE
tls: optionally allow expired TLS client certs.

### PR DESCRIPTION
**Pull request by doughd**
*Description*: In attempting to replace HAProxy with Envoy, we need to be able to handle expired certs.  We want to require all connections have certs signed by our CA and any expired certs (that are signed by our CA) are redirected to another route.  The redirection is going to happen through a filter, but we need a way to let Envoy accept the connections.  Currently, Envoy enforces `NotBefore` and `NotAfter` so it will reject these connections.  Even if you leave `require_client_certificate` to default false, it will still reject expired certs when presented.  With this option enabled, it will allow expired certs.

In HAProxy, this functionality corresponds to:

```
use_backend expiredcert if { ssl_c_verify 10 }
```

or

```
redirect location ... if ...
```

*Risk Level*: Medium risk as the default is still to enable verification.  High risk if the logic is wrong.

*Testing*: Added test cases in this PR for the default case, explicitly allowing and explicitly denying cert expiration verification.  I also did manual testing with an envoy + golang backend + grpcurl client setup (since the latter can handle TLS certs with gRPC connections).

*Docs Changes*: N/A other than a new option added to the protos.

*Release Notes*: Added a note to the version history.

### CODE REVIEW

**Comment by mattklein123** [comment]
@doughd thanks for the contribution. @ggreenway @PiotrSikora can you take a first pass / sanity check on this?

File: source/common/config/tls_context_json.cc [hunk 1]
```diff
@@ -59,6 +59,10 @@ void TlsContextJson::translateCommonTlsContext(
   for (const auto& san : json_tls_context.getStringArray("verify_subject_alt_name", true)) {
     validation_context->add_verify_subject_alt_name(san);
   }
+  if (json_tls_context.hasObject("verify_certificate_expiration")) {
```
Comment (mattklein123) [inline source/common/config/tls_context_json.cc]: Drive by comment: Let's please not add this feature to v1 config. Can we revert this change and the schema change? Thank you.

File: api/envoy/api/v2/auth/cert.proto [hunk 2]
```diff
@@ -171,6 +171,9 @@ message CertificateValidationContext {
   // certificate has not been revoked by this CRL. If this DataSource contains
   // multiple CRLs, all of them will be used.
   core.DataSource crl = 7;
+
+  // If specified, Envoy will not reject expired certs.
+  google.protobuf.BoolValue verify_certificate_expiration = 8;
```
Comment (ggreenway) [inline api/envoy/api/v2/auth/cert.proto]: Please use bool instead of google.protobuf.BoolValue, and rename so that false is the default, eg "disable_certificate_expiration_verification" or similar.

File: source/common/config/tls_context_json.cc (hunk 1 above)
Comment (ggreenway) [inline source/common/config/tls_context_json.cc]: We're trying not to add any more v1 config; it is deprecated. Please revert this, and the schema change.

File: source/common/ssl/context_impl.cc [hunk 3]
```diff
@@ -131,6 +131,15 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
...
   }
 
+  // NOTE: We're using SSL_CTX_set_cert_verify_callback() instead of X509_verify_cert()
+  // directly. However, our new callback is still calling X509_verify_cert() under
+  // the hood. Therefore, to ignore cert expiration, we need to set the callback
+  // for X509_verify_cert to ignore that error.
+  if (config.disableCertificateExpirationVerification()) {
+    X509_STORE* store = SSL_CTX_get_cert_store(ctx_.get());
+    X509_STORE_set_verify_cb(store, ContextImpl::verifyIgnoreCertExpirationCallback);
+  }
```
[the remark this answers is not in the document]
Comment (doughd) [reply source/common/ssl/context_impl.cc]: It's ugly, but I think necessary since `ContextImpl::verifyCallback()` is piggybacking on `X509_verify_cert()` for part of the validation.

`X509_verify_cert` calls `ctx->verify` (or the default `internal_verify`) which loops through the chain.  If an iteration fails due to an expiration in `internal_verify`, it will abort the loop and exit early which then cascades as `X509_verify_cert` also exits early on error.  If you check for expiration as an error condition from `X509_verify_cert` rather than using a callback to ignore it, you're skipping over other validation checks.

File: source/common/ssl/context_impl.cc [hunk 4]
```diff
@@ -241,6 +250,18 @@ bssl::UniquePtr<SSL> ContextImpl::newSsl() const {
   return bssl::UniquePtr<SSL>(SSL_new(ctx_.get()));
 }
 
+int ContextImpl::verifyIgnoreCertExpirationCallback(int ok, X509_STORE_CTX* ctx) {
+  int err = X509_STORE_CTX_get_error(ctx);
+
+  if (err == X509_V_OK || err == X509_V_ERR_CERT_HAS_EXPIRED ||
+      err == X509_V_ERR_CERT_NOT_YET_VALID) {
+    return 1;
```
[the remark this answers is not in the document]
Comment (doughd) [reply source/common/ssl/context_impl.cc]: Hmm I'm not sure how I would get access to `stats_`.  Unfortunately, I'm not passed in a `void *` that I could cast to an object like `verifyCallback` does.  All I have is an `int` and `X509_STORE_CTX *`.  There is a `void *other_ctx` in there but it's used by `get_issuer`.  IIRC I can't make it a lambda with capture parameters and still be compatible.

File: test/common/ssl/ssl_socket_test.cc [hunk 5]
```diff
@@ -678,6 +678,89 @@ TEST_P(SslSocketTest, FailedClientAuthSanVerification) {
...
+  client_cert->mutable_certificate_chain()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_cert.pem"));
+  client_cert->mutable_private_key()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_key.pem"));
+
+  // ssl.handshake logged by both: client & server.
+  testUtilV2(listener, client, "", true, "", "", "spiffe://lyft.com/test-team", "", "ssl.handshake",
+             2, GetParam());
+}
+
```
[the remark this answers is not in the document]
Comment (doughd) [reply test/common/ssl/ssl_socket_test.cc]: Ok, I checked the wrong CA manually.  I'll add those tests on Monday.

File: source/common/ssl/context_impl.cc [hunk 6]
```diff
@@ -253,12 +253,13 @@ bssl::UniquePtr<SSL> ContextImpl::newSsl() const {
 int ContextImpl::verifyIgnoreCertExpirationCallback(int ok, X509_STORE_CTX* ctx) {
   int err = X509_STORE_CTX_get_error(ctx);
```
Comment (PiotrSikora) [inline source/common/ssl/context_impl.cc]: This should be inside `(!ok)` block, since that's an extra call that's unused outside of it.

File: source/common/ssl/context_impl.h [hunk 7]
```diff
@@ -84,10 +84,10 @@ class ContextImpl : public virtual Context {
    */
   static int sslContextIndex();
 
-  // A X509_STORE_set_verify_cb callback for a custom error handling by X509_verify_cert().
+  // A X509_STORE_CTX_verify_cb callback for ignoring cert expiration in X509_verify_cert().
   static int verifyIgnoreCertExpirationCallback(int ok, X509_STORE_CTX* store_ctx);
```
Comment (PiotrSikora) [inline source/common/ssl/context_impl.h]: Nit: this should probably be called `ignoreCertificateExpirationCallback`.

File: docs/root/intro/version_history.rst [hunk 8]
```diff
@@ -108,6 +108,8 @@ Version history
   is supplied with the client request, its value will override any sampling decision made by the Envoy proxy.
 * websocket: support configuring
   :ref:`idle_timeout and max_connect_attempts <envoy_api_field_route.RouteAction.websocket_config>`.
+* tls: added support for allowing expired certificates with
```
Comment (PiotrSikora) [inline docs/root/intro/version_history.rst]: Nit: release notes is configured alphabetically, this should probably be added just before `tls: removed...`.

File: source/common/ssl/context_impl.cc [hunk 9]
```diff
@@ -131,6 +131,15 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
     verify_mode = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
   }
 
+  // NOTE: We're using SSL_CTX_set_cert_verify_callback() instead of X509_verify_cert()
+  // directly. However, our new callback is still calling X509_verify_cert() under
+  // the hood. Therefore, to ignore cert expiration, we need to set the callback
+  // for X509_verify_cert to ignore that error.
+  if (config.allowExpiredCertificate()) {
```
Comment (PiotrSikora) [inline source/common/ssl/context_impl.cc]: Actually, this should be inside the `if (!config.caCert().empty())` block, since this makes sense only when `trusted_ca` is configured (see: https://github.com/envoyproxy/envoy/pull/3549, https://github.com/envoyproxy/envoy/pull/3550).

Comment (doughd) [reply source/common/ssl/context_impl.cc]: Doesn't it also make sense in the `verify_certificate_hash` case without a CA per #3549 ?

How about if I move it into the`if (verify_mode != SSL_VERIFY_NONE)` block where it sets `verifyCallback`?

Comment (PiotrSikora) [reply source/common/ssl/context_impl.cc]: No, it doesn't, since validity period is a property of CA's signature.

Both: `verify_certificate_spki` and `verify_certificate_hash` verify only hash of SPKI / certificate, when used without `trusted_ca`.

File: docs/root/intro/version_history.rst [hunk 10]
```diff
@@ -127,6 +127,8 @@ Version history
 * stats: added support for histograms.
 * stats: added :ref:`option to configure the statsd prefix<envoy_api_field_config.metrics.v2.StatsdSink.prefix>`
 * stats: updated stats sink interface to flush through a single call.
+* tls: added support for allowing expired certificates with
```
Comment (PiotrSikora) [inline docs/root/intro/version_history.rst]: Nit: could you move it to after `verify_certificate_*` entries but before `allow_renegotiation`?

File: source/common/ssl/context_config_impl.cc [hunk 11]
```diff
@@ -74,8 +75,10 @@ ContextConfigImpl::ContextConfigImpl(const envoy::api::v2::auth::CommonTlsContex
     }
     if (!verify_subject_alt_name_list_.empty()) {
       throw EnvoyException(fmt::format("SAN-based verification of peer certificates without "
-                                       "trusted CA is insecure and not allowed",
-                                       certificateRevocationListPath()));
+                                       "trusted CA is insecure and not allowed"));
+    }
+    if (allow_expired_certificate_) {
+      throw EnvoyException(fmt::format("Cannot ignore certificate expiration without trusted CA"));
```
Comment (PiotrSikora) [inline source/common/ssl/context_config_impl.cc]: Maybe something like `"Certificate validity period is always ignored without trusted CA"` would be a bit more accurate? What do you think?

File: source/common/ssl/context_impl.cc [hunk 12]
```diff
@@ -146,6 +146,16 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
 
   if (verify_mode != SSL_VERIFY_NONE) {
     SSL_CTX_set_verify(ctx_.get(), verify_mode, nullptr);
+
+    // NOTE: We're using SSL_CTX_set_cert_verify_callback() instead of X509_verify_cert()
```
Comment (PiotrSikora) [inline source/common/ssl/context_impl.cc]: Looks like you moved this out from the `if (!caCert.empty())` block in the merge commit? Any reason to do that? Like I said previously, it's useless there, since `X509_verify_cert()` won't be called without trusted CA.

File: docs/root/intro/version_history.rst (hunk 10 above)
Comment (doughd) [reply docs/root/intro/version_history.rst]: Sure

File: source/common/ssl/context_config_impl.cc (hunk 11 above)
Comment (doughd) [reply source/common/ssl/context_config_impl.cc]: Works for me.

File: source/common/ssl/context_impl.cc (hunk 12 above)
Comment (doughd) [reply source/common/ssl/context_impl.cc]: Ok, I'll move it back.  I was trying to put both verify callback configurations in one place.  It seems equivalent since `if (!caCert.empty())` sets `verify_trusted_ca_` and `verify_mode`.  After the merge, `verify_trusted_ca_` was available.

File: api/envoy/api/v2/auth/cert.proto [hunk 13]
```diff
@@ -213,6 +213,9 @@ message CertificateValidationContext {
   // certificate has not been revoked by this CRL. If this DataSource contains
   // multiple CRLs, all of them will be used.
   core.DataSource crl = 7;
+
+  // If specified, Envoy will not reject expired certs.
```
Comment (PiotrSikora) [inline api/envoy/api/v2/auth/cert.proto]: Nit: `s/certs/certificates/`

File: source/common/ssl/context_impl.cc [hunk 14]
```diff
@@ -254,6 +262,18 @@ bssl::UniquePtr<SSL> ContextImpl::newSsl() const {
...
 
+int ContextImpl::ignoreCertificateExpirationCallback(int ok, X509_STORE_CTX* ctx) {
+  if (!ok) {
+    int err = X509_STORE_CTX_get_error(ctx);
+    if (err == X509_V_ERR_CERT_HAS_EXPIRED || err == X509_V_ERR_CERT_NOT_YET_VALID) {
+      return 1;
+    }
+  }
+
+  // Otherwise, return the original value
```
Comment (PiotrSikora) [inline source/common/ssl/context_impl.cc]: Nit: I don't think this comment adds anything here, so let's just removed it.

File: test/common/ssl/ssl_socket_test.cc [hunk 15]
```diff
@@ -678,6 +678,159 @@ TEST_P(SslSocketTest, FailedClientAuthSanVerification) {
            GetParam());
 }
 
+TEST_P(SslSocketTest, FailedClientCertificateDefaultExpirationVerification) {
```
Comment (ggreenway) [inline test/common/ssl/ssl_socket_test.cc]: Please add a sentence or two comment before each test function explaining what it is testing. Many existing tests don't have this, but we're trying to add it for new tests.

File: test/common/ssl/ssl_socket_test.cc [hunk 16]
```diff
@@ -678,6 +678,159 @@ TEST_P(SslSocketTest, FailedClientAuthSanVerification) {
...
+  server_cert->mutable_private_key()->set_filename(
+      TestEnvironment::substitute("{{ test_tmpdir }}/unittestkey.pem"));
+  envoy::api::v2::auth::CertificateValidationContext* server_validation_ctx =
+      filter_chain->mutable_tls_context()
+          ->mutable_common_tls_context()
+          ->mutable_validation_context();
+  server_validation_ctx->mutable_trusted_ca()->set_filename(
+      TestEnvironment::substitute("{{ test_rundir }}/test/common/ssl/test_data/ca_cert.pem"));
+
+  // Enforce expiration checks (cheked by default as well).
```
Comment (ggreenway) [inline test/common/ssl/ssl_socket_test.cc]: spelling: checked

File: test/common/ssl/ssl_socket_test.cc [hunk 17]
```diff
@@ -678,6 +678,159 @@ TEST_P(SslSocketTest, FailedClientAuthSanVerification) {
...
+  client_cert->mutable_certificate_chain()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_cert.pem"));
+  client_cert->mutable_private_key()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_key.pem"));
+
+  testUtilV2(listener, client, "", false, "", "", "spiffe://lyft.com/test-team", "",
+             "ssl.fail_verify_error", 1, GetParam());
+}
+
+TEST_P(SslSocketTest, ClientCertificateExpirationAllowedVerification) {
```
Comment (ggreenway) [inline test/common/ssl/ssl_socket_test.cc]: I didn't look too carefully, but it looks like this test is identical to the previous one, except the value for set_allow_expired_certificate and the expected result. Can the two share a single function taking a boolean? If it's messy, don't worry about it.

Comment (doughd) [reply test/common/ssl/ssl_socket_test.cc]: They could share a single function taking a boolean if it was part of `SslSocketTest` because it needs access to `GetParam()`.  Seems a bit messy when only two methods need this.

However, I could have a utility function so it's reduced to:

```
TEST_P(SslSocketTest, FailedClientCertificateExpirationVerification) {
  envoy::api::v2::Listener listener;
  envoy::api::v2::auth::UpstreamTlsContext client;

  configureExpiredClientCertificate(listener, client, false);

  testUtilV2(listener, client, "", false, "", "", "spiffe://lyft.com/test-team", "",
             "ssl.fail_verify_error", 1, GetParam());
}
```
